tenseal/cpp/tensors/utils/matrix_ops.h

```diff
@@ -149,21 +102,29 @@ Ciphertext diagonal_ct_vector_matmul_parallel(
             tenseal_context->evaluator->rotate_vector_inplace(
                 ct, local_i, *tenseal_context->galois_keys());
 
-            // accumulate results
-            result_mutex.lock();
-            tenseal_context->evaluator->add_inplace(result, ct);
-            result_mutex.unlock();
+            // accumulate thread results
+            tenseal_context->evaluator->add_inplace(thread_result, ct);
         }
+        return thread_result;
     };
 
-    // if not specified (n_threads set to 0), detect automatically
-    if (n_threads == 0) n_threads = get_concurrency();
-    vector<thread> threads;
-    threads.reserve(n_threads);
-    // start the threads
-    for (uint i = 0; i < n_threads; i++) threads.push_back(thread(thread_func));
-    // wait for the threads
-    for (uint i = 0; i < n_threads; i++) threads[i].join();
+    if (tenseal_context->get_concurrency() == 1)
+        return worker_func(0, vector_size);
+
+    size_t thread_cnt = tenseal_context->get_concurrency();
+    std::vector<std::future<Ciphertext>> future_results;
+    size_t batch_size = (vector_size + thread_cnt - 1) / thread_cnt;
+
+    for (size_t i = 0; i < thread_cnt; i++) {
+        future_results.push_back(tenseal_context->dispatcher()->enqueue_task(
+            worker_func, i * batch_size,
+            std::min((i + 1) * batch_size, vector_size)));
+    }
+
+    for (size_t i = 0; i < tenseal_context->get_concurrency(); i++) {
+        tenseal_context->evaluator->add_inplace(result,
+                                                future_results[i].get());
+    }
```

Comment:
So, you're saying that this loop might get stuck at 0, let's say, waiting for a slow thread, while the others finished the job, right?

Windows has a mechanism for waiting for multiple objects efficiently, unfortunately I cannot find something similar in standard C++ library, but I'll research it. I was looking for a solution too here, it's not ideal to loop over the futures and do a blocking wait. 

I could experiment with more designs. One other idea was to add the results to a blocking queue, and notify the main thread using a condition variable. That might schedule the operations more efficiently.

Reply:
The current design is great, I don't think we should jump directly into a different one, I just wanted to give it some thoughts. But this is definitely something I would merge

Reply:
I'll make a test with a condition variable&mutex, to see how the benchmarks look that way.

Reply:
I did a test using a mutex and a condition variable to notify the main thread(instead of waiting for a std::future), but the benchmarks look pretty similar.

I think it needs better load testing here, I am not sure how to simulate the scenario.

Maybe a random sleep in each worker might show the performance differences(cond variable vs future.get() )
I have the git stash anyway, I can try to add a test in another PR.